server/src/main/java/org/elasticsearch/index/mapper/DocumentMapperParser.java

```diff
@@ -22,57 +22,42 @@
-    private final SimilarityService similarityService;
-    private final Supplier<QueryShardContext> queryShardContextSupplier;
+    private final IndexSettings indexSettings;
+    private final IndexAnalyzers indexAnalyzers;
+    private final Function<String, String> documentTypeResolver;
+    private final DocumentParser documentParser;
+    private final Supplier<Map<Class<? extends MetadataFieldMapper>, MetadataFieldMapper>> metadataMappersSupplier;
+    private final Supplier<Mapper.TypeParser.ParserContext> parserContextSupplier;
     private final RootObjectMapper.TypeParser rootObjectTypeParser = new RootObjectMapper.TypeParser();
-    private final Version indexVersionCreated;
-    private final Map<String, Mapper.TypeParser> typeParsers;
     private final Map<String, MetadataFieldMapper.TypeParser> rootTypeParsers;
-    private final ScriptService scriptService;
-
-    public DocumentMapperParser(IndexSettings indexSettings,
-                                MapperService mapperService,
-                                SimilarityService similarityService,
-                                MapperRegistry mapperRegistry,
-                                Supplier<QueryShardContext> queryShardContextSupplier,
-                                ScriptService scriptService) {
-        this.mapperService = mapperService;
-        this.similarityService = similarityService;
-        this.queryShardContextSupplier = queryShardContextSupplier;
-        this.scriptService = scriptService;
-        this.typeParsers = mapperRegistry.getMapperParsers();
-        this.indexVersionCreated = indexSettings.getIndexVersionCreated();
-        this.rootTypeParsers = mapperRegistry.getMetadataMapperParsers(indexVersionCreated);
-    }
-
-    public Mapper.TypeParser.ParserContext parserContext() {
-        return new Mapper.TypeParser.ParserContext(similarityService::getSimilarity, typeParsers::get, indexVersionCreated,
-            queryShardContextSupplier, null, scriptService, mapperService.getIndexAnalyzers(), mapperService.getIndexSettings(),
-            mapperService::isIdFieldDataEnabled);
-    }
 
-    public Mapper.TypeParser.ParserContext parserContext(DateFormatter dateFormatter) {
-        return new Mapper.TypeParser.ParserContext(similarityService::getSimilarity, typeParsers::get, indexVersionCreated,
-            queryShardContextSupplier, dateFormatter, scriptService, mapperService.getIndexAnalyzers(), mapperService.getIndexSettings(),
-            mapperService::isIdFieldDataEnabled);
+    DocumentMapperParser(IndexSettings indexSettings,
+                         IndexAnalyzers indexAnalyzers,
+                         Function<String, String> documentTypeResolver,
+                         DocumentParser documentParser,
+                         Supplier<Map<Class<? extends MetadataFieldMapper>, MetadataFieldMapper>> metadataMappersSupplier,
+                         Supplier<Mapper.TypeParser.ParserContext> parserContextSupplier,
```

Comment:
This is not perfect, but it gets the job done. The need for these two suppliers and a function here is a bad signal I think, but it's the best I could come up with for now.

Reply:
Does the metadata mappers map need to be a Supplier?

Reply:
this annoys me too. I believe it has to, because the doc mapper can change in MapperService?